src/Bedrock.Framework/Protocols/ConsumableArrayBufferWriter.cs

```diff
@@ -0,0 +1,235 @@
+
+        /// <summary>
+        /// Returns a <see cref="Span{T}"/> to write to that is at least the requested length (specified by <paramref name="sizeHint"/>).
+        /// If no <paramref name="sizeHint"/> is provided (or it's equal to <code>0</code>), some non-empty buffer is returned.
+        /// </summary>
+        /// <exception cref="ArgumentException">
+        /// Thrown when <paramref name="sizeHint"/> is negative.
+        /// </exception>
+        /// <remarks>
+        /// This will never return an empty <see cref="Span{T}"/>.
+        /// </remarks>
+        /// <remarks>
+        /// There is no guarantee that successive calls will return the same buffer or the same-sized buffer.
+        /// </remarks>
+        /// <remarks>
+        /// You must request a new buffer after calling Advance to continue writing more data and cannot write to a previously acquired buffer.
+        /// </remarks>
+        public Span<T> GetSpan(int sizeHint = 0)
+        {
+            CheckAndResizeBuffer(sizeHint);
+            Debug.Assert(_buffer.Length > _index);
+            return _buffer.AsSpan(_index);
+        }
+
+        private void CheckAndResizeBuffer(int sizeHint)
+        {
+            if (sizeHint < 0)
+                throw new ArgumentException(nameof(sizeHint));
+
+            if (sizeHint == 0)
+            {
+                sizeHint = 1;
+            }
+
+
+            if (sizeHint > FreeCapacity)
+            {
+                var countUnconsumed = _index - _consumedCount;
+                var countAvailable = _buffer.Length - countUnconsumed;
+                // If doing so would give us significant free capacity (half the array arbitrarily chosen here)
+                // Shift the array left rather than allocating a new array.
+                if (countAvailable > sizeHint&& countAvailable > _buffer.Length / 2)
+                {
+                    Array.Copy(_buffer, _consumedCount, _buffer, 0, countUnconsumed);
+                }
+                else
+                {
+                    var growBy = Math.Max(sizeHint - _consumedCount, _buffer.Length);
+
+                    if (_buffer.Length == 0)
+                    {
+                        growBy = Math.Max(growBy, DefaultInitialBufferSize);
+                    }
+
+                    var newSize = checked(_buffer.Length + growBy);
+                    var destinationArray = new T[newSize];
+                    Array.Copy(_buffer, _consumedCount, destinationArray, 0, countUnconsumed);
+                    _buffer = destinationArray;
+                }
+                _index = countUnconsumed;
+                _consumedCount = 0;
+            }
+
+            Debug.Assert(FreeCapacity > 0 && FreeCapacity >= sizeHint);
+        }
+
+        private static void ThrowInvalidOperationException_AdvancedTooFar(int capacity)
+        {
+            throw new InvalidOperationException($"BufferWriter advanced too far. Capacity: {capacity}");
+        }
+    }
+}
```

Comment:
Nit: newline

Reply:
You mean  you want a newline at the end of the file? Is that the style here?

Reply:
Yes.
Right now no explicit style is defined, but @davidfowl has this on his list (as he said in other comments) and will update the editorconfig.
Newlines at the end is pretty much everywhere in the .NET-landscape (and beyond).